### PR TITLE
Submit php-rpc-lib milestone 1

### PR DESCRIPTION
**Pull request by gmajor-encrypt**
# Milestone Delivery Checklist

- [x] The [milestone-delivery-template.md](https://github.com/w3f/Grant-Milestone-Delivery/blob/master/deliveries/milestone-delivery-template.md) has been copied and updated.
- [x] The [invoice form :pencil:](https://forms.gle/LSRr7PCjBpEbKGh89) has been filled out for this milestone.
- [x] This pull request is being made by the same account as the accepted application.
- [x] In case of acceptance, the payment will be transferred to the BTC/ETH payment address provided in the application.
- [x] The delivery is according to the [Guidelines for Milestone Deliverables](https://github.com/w3f/Grants-Program/blob/master/docs/milestone-deliverables-guidelines.md).

Link to the application pull request: https://github.com/w3f/Grants-Program/pull/622

### CODE REVIEW

**Comment by alxs** [comment]
Thanks for the delivery @gmajor-encrypt, we'll look into it as soon as possible.

File: deliveries/php-rpc-lib_milestone_1.md [hunk 1]
```diff
@@ -0,0 +1,20 @@
+# Milestone Delivery :mailbox:
+
+
+**The [invoice form :pencil:](https://docs.google.com/forms/d/e/1FAIpQLSfmNYaoCgrxyhzgoKQ0ynQvnNRoTmgApz9NrMp-hd8mhIiO0A/viewform) has been filled out correctly for this milestone.**
+
+* **PR Link:** [php-rpc-lib](https://github.com/w3f/Grants-Program/pull/622)
```
Comment (alxs) [inline deliveries/php-rpc-lib_milestone_1.md:6]: We updated the [template](https://github.com/w3f/Grant-Milestone-Delivery/blob/master/deliveries/milestone-delivery-template.md) some time ago, this will get the checks to pass.

```suggestion
* **Application Document:** https://github.com/w3f/Grants-Program/blob/master/applications/php-rpc-lib.md
```

**Comment by alxs** [comment]
Hey @gmajor-encrypt, sorry for the late reply. I have mostly the same feedback as for previous deliveries (: Documentation and tests are far from comprehensive - familiarising yourself with a few basic testing principles would go a long way towards improving the quality of the project and in turn people's readiness to trust your code.

For example, I have no idea what [this test](https://github.com/gmajor-encrypt/php-substrate-api/blob/master/test/Rpc/ClientTest.php#L18) does. It has a vague name, no description, tests at least two things (connection and then something (?) that is expected to raise an exception). And why does it use an http URL for a variable named `wsClient`? I really don't know what I'm looking at, and I wouldn't let this code anywhere near production.

I'd also expect to see different cases testing edge cases, corner cases, exceptions, etc. This goes for both repositories, please see what you can do.

As for the documentation, please describe and provide instructions for the full API.

**Comment by gmajor-encrypt** [comment]
@alxs Thank you for your review, sorry for the late reply, China is celebrating the Spring Festival these days, and changes will be made in the near future

**Comment by gmajor-encrypt** [comment]
@alxs Sorry for the wait. I improved the readme doc and test case, thank you for your review

**Comment by alxs** [comment]
Thanks a lot for the extra effort. The documentation looks great, now I found it much easier to understand the API. The tests also look good, though some of them in `php-substrate-api` are [failing](https://github.com/w3f/Grant-Milestone-Delivery/blob/master/evaluations/php-rpc-lib_1_alxs.md#general-notes). Could you look into it? It looks like one needs to build the go files for the bindings, but that probably won't be possible when using composer?

**Comment by gmajor-encrypt** [comment]
@alxs Hi, The current default sr25519 FFI is for mac. If it is not for mac, it needs to be recompiled. Unfortunately, PHP composer currently does not support automatic compilation after installation, so manual compilation is required. https://github.com/gmajor-encrypt/php-substrate-api/blob/master/.github/workflows/ci.yml#L42 Here you can see the compilation process. I will improve this issue in the documentation

**Comment by gmajor-encrypt** [comment]
@alxs  Readme has been updated. https://github.com/gmajor-encrypt/php-substrate-api#troubleshooting

**Comment by alxs** [comment]
Ok, thanks. Could you push the new version with the instructions to packagist?

By the way, a convenience function for the below preliminary boilerplate required to read from state would likely be appreciated:

```php
$codec = new ScaleInstance(Base::create());
$metadataV14RawValue = "...." //  from json rpc state_getMetadata
$metadata = $codec->process("metadata", new ScaleBytes($metadataV14RawValue))["metadata"];
```

**Comment by gmajor-encrypt** [comment]
@alxs  packagist will automatically update immediately after push.

**Comment by alxs** [comment]
Perfect. In that case, I'm happy to let you know that the milestone is a pass! You can find my evaluation notes [here](https://github.com/w3f/Grant-Milestone-Delivery/blob/master/evaluations/php-rpc-lib_1_alxs.md#general-notes). I'll notify the operations team to pay out your invoice for this milestone.

Good luck with the next milestone and as usual, please let us know if you need anything or run into delays.

**Comment by Noc2** [comment]
@gmajor-encrypt Our invoice team just pointed out that a few information are missing on your invoice: Name, address, post code and place. Could you add this information and send us a new invoice? Thanks.

**Comment by gmajor-encrypt** [comment]
@Noc2 hi, I just submitted the invoice through https://forms.gle/LSRr7PCjBpEbKGh89, please help to check

**Comment by Noc2** [comment]
Perfect. I will forward it internally

**Comment by RouvenP** [comment]
hi @gmajor-encrypt we transferred the payment today. Thanks!
